# An Application of Eigenvalues and Eigenvectors 

A car rental company has offices in Manhattan and the Bronx. Relying on its records, the company knows that on a monthly basis $40\% $ of rentals from the Manhattan office are returned there and $60\%$ are one-way rentals that are dropped off in the Bronx. Similarly, $70\%$ of rentals from the Bronx office are returned there, whereas $30\%$ are dropped off in Manhattan. 

Let $m_k$ and $b_k$ denote the number of cars at the depots in Manhattan and the Bronx, respectively, at the beginning of month $k$ ($k = 0, 1, 2, ...$). One month later, the cars at the Manhattan location consist of those returned there during the previous month (i.e., $40\%$ of $m_k$), together with those dropped off on a one-way rental from the Bronx office (i.e. $30 \%$ of $b_k$). Similarly, the cars at the Bronx location consist of those returned there during the previous month (i.e., $70\%$ of $b_k$), together with those dropped off on a one-way rental from the Manhattan office (i.e. $60\%$ of $m_k$). We can express this information in terms of _difference equations_ (which are analogous to differential equations): 

$$ 
\begin{cases}
    m_{k+1} = 0.4m_k + 0.3b_k\\
    b_{k+1} = 0.6m_k + 0.7b_k 
\end{cases}

\hspace{1cm}
k = 0, 1, 2, ...
$$

The matrix-vector form of this system is: 

$$
\begin{bmatrix}
m_{k+1} \\
b_{k+1} \\ 
\end{bmatrix} = 

\begin{bmatrix}
0.4 & 0.3 \\
0.6 & 0.7 \\ 
\end{bmatrix}


\begin{bmatrix}
m_k \\
b_k \\ 
\end{bmatrix}

\hspace{1cm}
k = 0, 1, 2, ...
$$



which we can write compactly as $X_{k+1} = AX_k$. We notice that 

$$
X_1 = A X_0 \hspace{1.3cm} \\
X_2 = A X_1 = A^2 X_0 \\
X_3 = A X_2 = A^3 X_0 \\ 
X_4 = A X_3 = A^4 X_0 \\ 
\vdots \hspace{1cm} \vdots \hspace{1.2cm}\\
X_k = A^k X_0 \hspace{1.2cm}
$$

Now, let's find the eigenvectors and eigenvalues of $A$: 

In [2]:
import numpy as np 

A = np.array([
    [0.4, 0.3], 
    [0.6, 0.7],
])

eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues: \n", eigenvalues, "\n") 
print("Eigenvectors: \n", eigenvectors)


Eigenvalues: 
 [0.1 1. ] 

Eigenvectors: 
 [[-0.70710678 -0.4472136 ]
 [ 0.70710678 -0.89442719]]


Thus the eigenpairs for $A$, are $(\lambda_1, v_1) = \Bigg(1, \begin{bmatrix}
-0.4472136\\ 
-0.89442719 \\ 
\end{bmatrix}\Bigg)$ and $(\lambda_2, v_2) = \Bigg(0.1, \begin{bmatrix}
-0.70710678\\ 
0.70710678 \\ 
\end{bmatrix}\Bigg)$

Since the eigenvectors $v_1$ and $v_2$ correspond to distinct eigenvalues of $A$, we know that they are linearly independent (by theorem 4.4.4). Thus the eigenvectors form a basis for $\mathbb{R}^2$, we can write any inital vector $X_0$ uniquely as $X_0 = c_1 v_1 + c_2 v_2$. 

$$
X_k = A^k X_0 = A^k (c_1 v_1 + c_2 v_2) = c_1 \lambda _1 ^ k v_1 + c_2 \lambda _2 ^ k v_2 = c_1 v_1 + c_2 (0.1)^k v_2
$$

Thus, 

$$
\lim_{k \rightarrow \infty} X_k = c_1 v_1
$$

In [5]:
v_1 = eigenvectors[:, 1]
print(f"v_1: {v_1}")
print(f"v_1 factored out: {v_1/v_1[0]}")


v_1: [-0.4472136  -0.89442719]
v_1 factored out: [1. 2.]


Thus, 

$$
\lim_{k \rightarrow \infty} X_k = c_1 v_1 =  -0.4472136 c_1 \begin{bmatrix}
1\\ 
2\\ 
\end{bmatrix} = c_3 \begin{bmatrix}
1\\ 
2\\ 
\end{bmatrix} =  \begin{bmatrix}
c_3\\ 
2 c_3\\ 
\end{bmatrix} = \lim_{k \rightarrow \infty} \begin{bmatrix}
m_k\\ 
b_k\\ 
\end{bmatrix}
$$

Thus, over time, the number of cars in the Manhattan depot tends to a value that's half the number of cars at the Bronx office. 

## Time for a simulation...

In [6]:
def simulate(manhattan_inital: int, bronx_inital: int, number_of_months: int): 
    m = manhattan_inital
    b = bronx_inital

    for _ in range(number_of_months): 
        m, b = 0.4 * m + 0.3 * b, 0.6 * m + 0.7 * b 
    
    print(f"Number of cars in Manhattan after {number_of_months} months: ", m)
    print(f"Number of cars in The Bronx after {number_of_months} months: ", b)


In [7]:
simulate(manhattan_inital=1000, bronx_inital=900, number_of_months=50)

Number of cars in Manhattan after 50 months:  633.3333333333329
Number of cars in The Bronx after 50 months:  1266.6666666666658


In [8]:
633 * 2 

1266

Thus our mathematical conclusion that the number of cars in The Bronx would be twice as much as the number of cars in Manhattan is correct (given enough time). 